feat(supernets): new local deployment process for v0.9 (#1626)
maticnetwork/matic-docs: docs/supernets/operate/local-supernet.md

- Generating private keys for PolyBFT nodes.
- Deploying and initializing rootchain contracts.
- Configuring the rootchain by allowlisting and registering validators, performing initial staking, and finalizing the validator set.
- Generating the genesis file and chain configuration.
- Funding validators on the rootchain.
- Running a (Supernet) cluster consisting of multiple PolyBFT nodes.

By the end of this tutorial, you will have a fully functional Supernet test network that can be used to process transactions with high throughput and low latency.

## What you'll do

This tutorial will cover the following steps:

1. [Node Preparation](#1-node-preparation)
   - i. [Generating account secrets](#i-generate-account-secrets)
   - ii. [Chain configuration: generating the Genesis file](#ii-chain-configuration-generate-the-genesis-file)
2. [Rootchain Configuration](#2-rootchain-configuration)
   - i. [Deploying rootchain contracts](#i-deploy-and-initialize-rootchain-contracts)
3. [Allowlisting](#3-allowlisting)

>    - i. [Allowlisting validators on the rootchain](#i-allowlist-validators-on-the-rootchain)

Reviewer: ```suggestion
   - i. [Whitelisting validators on the rootchain](#i-whitelist-validators-on-the-rootchain)
```

   Multi-host:

   ```bash
   ./polygon-edge genesis --block-gas-limit 10000000 --epoch-size 10 --validators /ip4/127.0.0.1/tcp/30301/p2p/16Uiu2HAmV5hqAp77untfJRorxqKmyUxgaVn8YHFjBJm9gKMms3mr:0xDcBe0024206ec42b0Ef4214Ac7B71aeae1A11af0:1cf134e02c6b2afb2ceda50bf2c9a01da367ac48f7783ee6c55444e1cab418ec0f52837b90a4d8cf944814073fc6f2bd96f35366a3846a8393e3cb0b19197cde23e2b40c6401fa27ff7d0c36779d9d097d1393cab6fc1d332f92fb3df850b78703b2989d567d1344e219f0667a1863f52f7663092276770cf513f9704b5351c4:11b18bde524f4b02258a8d196b687f8d8e9490d536718666dc7babca14eccb631c238fb79aa2b44a5a4dceccad2dd797f537008dda185d952226a814c1acf7c2]
   ```

3. Deploy and initialize rootchain contracts:

   [FOR GETH ONLY] Start rootchain server:

   ```bash
   ./polygon-edge rootchain server
   ```

   ```bash
   polygon-edge rootchain deploy --deployer-key <hex_encoded_rootchain_account_private_key> [--genesis ./genesis.json] [--json-rpc http://127.0.0.1:8545] [--test]
   ```

4. Fund validators on rootchain:

   ```bash
   polygon-edge rootchain fund --data-dir ./test-chain-1
   ```

> 5. Allowlist validators on rootchain:

Reviewer: ```suggestion
5. Whitelist validators on rootchain:
```

    ```bash
    ./polygon-edge server --data-dir ./test-chain-1 --chain genesis.json --grpc-address :5001 --libp2p :30301 --jsonrpc :9545 \
    --seal --log-level DEBUG --relayer
    ```

</details>

## 1. Node preparation

### i. Generating account secrets

To initialize PolyBFT consensus, we need to generate the necessary secrets for each node.

The `polygon-edge polybft-secrets` command is used to generate account secrets for validators. The command initializes private keys for the consensus client (validators + networking) to a Secrets Manager config file.

**Keep in mind that this command is intended for testing purposes only and should not be used in a production environment.**

<details>
<summary>Flags</summary>

| Flag            | Description                                                                                               | Example           |
|-----------------|-----------------------------------------------------------------------------------------------------------|-------------------|
| `--account`     | Indicates whether a new account should be created (default true).                                         |                   |

> | `--chain-id`    | Specifies the ID of the chain (default 100).                                                              | `--chain-id 333`  |

Reviewer: We don't have a default `chain-id` anymore, since `chain-id` is provided by `StakeManager` on deployment of rootchain contracts (in the `deploy` command).

Author: Updated.

1. [Node Preparation](#1-node-preparation)
   - i. [Generating account secrets](#i-generate-account-secrets)
   - ii. [Chain configuration: generating the Genesis file](#ii-chain-configuration-generate-the-genesis-file)
2. [Rootchain Configuration](#2-rootchain-configuration)
   - i. [Deploying rootchain contracts](#i-deploy-and-initialize-rootchain-contracts)
3. [Allowlisting](#3-allowlisting)
   - i. [Allowlisting validators on the rootchain](#i-allowlist-validators-on-the-rootchain)
   - ii. [Register validators on the rootchain](#ii-register-validators-on-the-rootchain)
4. [Staking](#4-staking)
   - i. [Initial staking on rootchain](#i-initial-staking-on-rootchain)
   - ii. [Finalizing validator set on the rootchain](#ii-finalize-validator-set-on-the-rootchain)
5. [Starting the Supernet](#5-start-the-supernet)

<details>
<summary>Fast-track guide</summary>

**Here's the fast-track guide if you're looking for a quick guide on the essential commands needed to set up a local Supernet.**

1. Init secrets:

   ```bash
   ./polygon-edge polybft-secrets --data-dir test-chain- --num 4
   ```

2. Create chain configuration:

   Single host:

   ```bash

>    polygon-edge genesis --block-gas-limit 10000000 --epoch-size 10 [--validators-path ./] [--validators-prefix test-chain-] [--consensus polybft] [--reward-wallet address:amount]

Reviewer 2: ```suggestion
   ./polygon-edge genesis --block-gas-limit 10000000 --epoch-size 10 [--validators-path ./] [--validators-prefix test-chain-] [--consensus polybft] [--reward-wallet address:amount]
```

5. Allowlist validators on rootchain:

   ```bash
   ./polygon-edge polybft whitelist-validators --private-key <hex_encoded_rootchain_account_private_key_of_supernetManager_deployer> --addresses <addresses_of_validators> --supernet-manager <address_of_SupernetManager_contract>
   ```

6. Register validators on rootchain:

   ```bash
   ./polygon-edge polybft register-validator --data-dir ./test-chain-1 --supernet-manager <address_of_SupernetManager_contract>
   ```

7. Initial staking on rootchain:

   ```bash
   ./polygon-edge polybft stake --data-dir ./test-chain-1 --chain-id <id_of_child_chain_from_genesis> --amount <amount_of_tokens_to_stake> --stake-manager <address_of_StakeManager_contract> --native-root-token <address_of_native_root_token>
   ```

8. Finalize genesis validator set on rootchain:

   ```bash
    ./polygon-edge polybft supernet --private-key <hex_encoded_rootchain_account_private_key_of_supernetManager_deployer> \
    --genesis <path_to_genesis_file> \
    --supernet-manager <address_of_SupernetManager_contract> \
    --stake-manager <address_of_StakeManager_contract> \
    --finalize-genesis --enable-staking
   ```

9. Run (child chain) cluster:

    ```bash

>     ./polygon-edge server --data-dir ./test-chain-1 --chain genesis.json --grpc-address :5001 --libp2p :30301 --jsonrpc :9545 \

Reviewer 2: Why `9545` and the rest on the 10k range? we should be consistent here

Author: Updated to use 10k range based on discussion. We'll revisit if needed.

> Note that the `--deployer-key` option is optional, and if you omit it, the default account in your local client will be used.

<details>
<summary>Flags</summary>

| Flag                  | Description                                                               | Example                                       |
|-----------------------|---------------------------------------------------------------------------|-----------------------------------------------|
| `--deployer-key`      | Hex encoded private key of the account which deploys rootchain contracts  | `--deployer-key <PRIVATE_KEY>`                |
| `--json-rpc`          | The JSON RPC rootchain IP address (e.g. http://127.0.0.1:8545)            | `--json-rpc http://127.0.0.1:8545`             |

> | `--path`              | Root directory path containing POS smart contracts                        | `--path ./core-contracts/artifacts/contracts/` |

Reviewer: I think we don't have this flag anymore since we use the `Go` generated artifacts of smart contracts that are in Edge codebase.

Author: Removed.

[ROOTCHAIN - CONTRACTS DEPLOYMENT] ExitHelper contract is initialized

[ROOTCHAIN - CONTRACTS DEPLOYMENT] RootERC20Predicate contract is initialized

[ROOTCHAIN - CONTRACTS DEPLOYMENT] RootERC721Predicate contract is initialized

[ROOTCHAIN - CONTRACTS DEPLOYMENT] RootERC1155Predicate contract is initialized

[ROOTCHAIN - CONTRACTS DEPLOYMENT] finished. All contracts are successfully deployed and initialized.
```

</details>

</TabItem>
<br/>

### ii. Funding validators on the rootchain

Before deploying validator nodes on the childchain network, we need to ensure that the validators have sufficient funds on the rootchain network. It's crucial to have enough funds in the validator account, as they need to cover the gas fees associated with their transactions on the rootchain.

To fund the validators' accounts on the rootchain, we use the `polygon-edge rootchain fund` command. When executed with the appropriate flags, it will:

1. Retrieve the validator account from the secrets manager.
2. Send a transaction to fund the validator account on the rootchain with a value of 10,000 $token (the network's native token).

> 3. Repeat steps 1 and 2 for each validator specified by the --num flag.

Reviewer: There is no `--num` flag anymore, it was removed from the command.

Author: Updated.

**It's important to note that this command is also for testing purposes only.**

In a production environment, you would need to ensure that the validators have sufficient funds on the rootchain network to cover the gas fees associated with their transactions.

Here's an example of how to fund each of the validator accounts:

  ```bash

>   ./polygon-edge rootchain fund --data-dir ./test-chain-1

Reviewer: I think we should add --amount flag to these examples, we don't provide the default amount in the command.

Author: Added.

  ./polygon-edge rootchain fund --data-dir ./test-chain-2
  ./polygon-edge rootchain fund --data-dir ./test-chain-3
  ./polygon-edge rootchain fund --data-dir ./test-chain-4
  ```

<details>
<summary>Funding output example</summary>

  ```bash
  [ROOTCHAIN FUND]
   Validator (address) = 0x61324166B0202DB1E7502924326262274Fa4358F
   Transaction (hash)  = 0x0fb6880751d3c2aa34e42680aa51e3848e2d8376960017068be1fe2fc9896786

  [ROOTCHAIN FUND]
   Validator (address) = 0xFE5E166BA5EA50c04fCa00b07b59966E6C2E9570
   Transaction (hash)  = 0x8e9890c55e391dc5e86e8edc271076039bb97d6ef9f5222b2e15fad73b7d7b87

   [ROOTCHAIN FUND]
   Validator (address) = 0x9aBb8441A12d4FD8D505C3fc50cDdc45E0df2b1e
   Transaction (hash)  = 0x9717b6feed8526d81227e690dc81d85eeb5b95580252b392d15372ec6c276447

   [ROOTCHAIN FUND]
   Validator (address) = 0xCaB5AAC79Bebe326e0c80d72b5662E73f5D8ea56
   Transaction (hash)  = 0xe3e35d80b19b61d8481d482bf9765769efb929de53d40a9d451f4efdcdd43bc0
  ```

</details>

> ## 3. Allowlisting

Reviewer: ```suggestion
## 3. Whitelisting
```
Let's keep it the same as in contracts and the name of the command.

Author: I understand your point about maintaining consistency, but I believe that using "allowlisting" instead of "whitelisting" is a better standard, which is also how it has been referred to (beyond the implementation.)

Author: Based on internal discussion, we'll keep as "Allowlist".

In this section, we will look at how to allowlist validators on the rootchain.

> ### i. Allowlist validators on the rootchain

Reviewer: ```suggestion
### i. Whitelist validators on the rootchain
```

The `CustomSupernetManager` contract on the rootchain is responsible for managing the PolyBFT network.

Before validators can be registered on the `CustomSupernetManager` contract on the rootchain, they need to be allowlisted by the deployer of the `CustomSupernetManager` contract. Validators can register themselves, or a registrator can register them on their behalf. Once registered, validators can stake and start validating transactions.

This can be done using the `polygon-edge polybft whitelist-validators` command. The deployer can specify the hex-encoded private key of the `CustomSupernetManager` contract deployer or use the `--data-dir` flag if they have initialized their secrets.

<details>
<summary>Flags</summary>

> 

Reviewer: Register command does not have `balance` flag anymore.

Author: It seems some commands are showing outdated flags. For instance:

```
polygon-edge polybft register-validator --help
Registers a new validator

Usage:
   polybft register-validator [flags]

Flags:
      --balance string                balance which is going to be funded to the new validator account (default "0xD3C21BCECCEDA1000000")
      --chain-id int                  the ID of the chain (default 100)
      --data-dir string               the directory path where new validator key is stored
  -h, --help                          help for register-validator
      --jsonrpc string                the JSON-RPC interface (default "0.0.0.0:8545")
      --registrator-data-dir string   the directory path where registrator validator key is stored
      --stake string                  stake represents amount which is going to be staked by the new validator account (default "0x56BC75E2D63100000")

Global Flags:
      --json   get all outputs in json format (default false)
```

Author: Updated.

> 

Reviewer: Register command does not have `chain-id` flag.

> 

Reviewer: We don't have this flag.

> 

Reviewer: We don't have this flag anymore.

> 

Reviewer: We don't have this flag.

Author: Updated.

> 

Reviewer: We should probably add `--supernet-manager` flag in this register command example. It's a flag that defines the address of the `CustomSupernetManager` contract on rootchain.

Author: Updated.

> 

Reviewer: We should probably add `--chain-id` , `--stake-manager`, `--native-root-token` flags for the stake command example.

Author: Updated.

> 

Reviewer: We don't have this flag anymore.

Author: Removed.

> 

Reviewer: We don't have this flag anymore.

Author: Removed.
